chef/cookbooks/utils/libraries/search.rb

```diff
@@ -48,5 +37,58 @@ def get_instance(query)
       end
       instance
     end
+
+    def node_search_with_cache(query, bc_instance = nil)
+      CrowbarUtilsSearch.node_search_with_cache(node, query, cookbook_name, bc_instance)
+    end
+  end
+end
+
+class CrowbarUtilsSearch
+  class << self
+    @node_search_cache = nil
+    @node_search_cache_time = nil
+
+    def node_search_with_cache(node, query, cookbook_name = nil, bc_instance = nil)
+      if @node_search_cache_time != node[:ohai_time]
+        Chef::Log.info("Invalidating node search cache") if @node_search_cache
+        @node_search_cache = {}
+        @node_search_cache_time = node[:ohai_time]
+      end
+
+      filter = crowbar_filter_env(node, query, cookbook_name, bc_instance)
+      real_query = "#{query}#{filter}"
+      @node_search_cache[real_query] ||= begin
+        results = Array.new
+        Chef::Search::Query.new.search(:node, real_query) do |o|
+          results << o
+        end
+        results
+      end
+    end
+
+    def crowbar_filter_env(node, query, cookbook_name = nil, bc_instance = nil)
+      # All cookbooks encode the barclamp name as the role name prefix, thus we can
+      # simply grab it from the query (e.g. BC 'keystone' for role 'keystone-server'):
+      return nil unless query =~ /^(roles|recipes):(\w*).*$/
+      barclamp = $2
+
+      # There are two conventions to filter by barclamp proposal:
+      #  1) Other barclamp cookbook: node[@cookbook_name][$OTHER_BC_NAME_instance]
+      #  2) Same cookbook: node[@cookbook_name][:config][:environment]
+      env = if !bc_instance.nil?
+        "#{barclamp}-config-#{bc_instance}"
+      elsif node[barclamp] && node[barclamp][:config] && (barclamp == cookbook_name)
+        node[barclamp][:config][:environment]
+      elsif !node[cookbook_name].nil? &&
+          !node[cookbook_name]["#{barclamp}_instance"].nil? &&
+          !node[cookbook_name]["#{barclamp}_instance"].empty?
+        "#{barclamp}-config-#{node[cookbook_name]["#{barclamp}_instance"]}"
+      end
```

Comment:
What will be the value for `env` if that failed on both `elif` tests? Will it be `nil` ?